File: F/1979-find-greatest-common-divisor-of-array/main.go
package main

import "fmt"

func findGCD(nums []int) int {
	min, max := nums[0], nums[0]
	for _, v := range nums {
		if min > v {
			min = v
		}
		if max < v {
			max = v
		}
	}

	var GCD func(x, y int) int
	GCD = func(x, y int) int {
		p1, p2 := x, y
		if p2 > p1 {
			p1, p2 = y, x
		}
		if p1%p2 == 0 {
			return p2
		}
		return GCD(p1%p2, p2)
	}

	return GCD(min, max)

}

func main() {
	fmt.Println()
}

File: F/1979-find-greatest-common-divisor-of-array/main.rs
impl Solution {
  pub fn find_gcd(nums: Vec<i32>) -> i32 {
    let gcd = |mut x: i32, mut y: i32| -> i32 {
      while y != 0 {
        let t = y;
        y = x % y;
        x = t;
      }
      x
    };

    gcd(*nums.iter().min().unwrap(), *nums.iter().max().unwrap())
  }

  pub fn find_gcd2(nums: Vec<i32>) -> i32 {
    let mut mn = nums[0].min(nums[1]);
    let mut mx = nums[0].max(nums[1]);

    (2 - nums.len() % 2..nums.len()).step_by(2).for_each(|idx| {
      let (mut x, mut y) = (nums[idx], nums[idx + 1]);
      if x > y {
        (x, y) = (y, x);
      }
      if x < mn {
        mn = x;
      }
      if y > mx {
        mx = y;
      }
    });

    let gcd = |mut x: i32, mut y: i32| -> i32 {
      while y != 0 {
        let t = y;
        y = x % y;
        x = t;
      }
      x
    };

    gcd(mn, mx)
  }
}
